tests/unit/h/views/api/groups_test.py

```diff
@@ -8,477 +8,606 @@
-            expand=["organization", "scopes"]
+        assert group_update_service.method_calls == [group_update_service_method_call]
+        assert_it_returns_group_as_json(
+            response, group=group_update_service.update.return_value
         )
-        assert result == GroupJSONPresenter.return_value.asdict.return_value
 
-    @pytest.fixture
-    def pyramid_request(self, pyramid_request, factories):
-        # Add a nominal json_body so that _json_payload() parsing of
-        # it doesn't raise
-        pyramid_request.json_body = {}
-        pyramid_request.user = factories.User()
-        return pyramid_request
+    def test_it_with_groupid_request_param(
+        self,
+        context,
+        pyramid_request,
+        group_service,
+        group_update_service,
+        UpdateGroupAPISchema,
+    ):
+        UpdateGroupAPISchema.return_value.validate.return_value["groupid"] = (
+            sentinel.groupid
+        )
+        group_service.fetch.return_value = context.group
 
-    @pytest.fixture(autouse=True)
-    def UpdateGroupAPISchema(self, patch):
-        return patch("h.views.api.groups.UpdateGroupAPISchema")
+        views.update(context, pyramid_request)
 
+        group_service.fetch.assert_called_once_with(pubid_or_groupid=sentinel.groupid)
+        assert group_update_service.update.call_args[1]["groupid"] == sentinel.groupid
 
-@pytest.mark.usefixtures(
-    "CreateGroupAPISchema", "group_service", "group_update_service"
-)
-class TestUpsertGroup:
-    def test_it_proxies_to_create_if_group_empty(
-        self, context, pyramid_request, groups_create
+    def test_it_with_duplicate_groupid(
+        self,
+        context,
+        pyramid_request,
+        group_service,
+        group_update_service,
+        UpdateGroupAPISchema,
     ):
-        context.group = None
+        UpdateGroupAPISchema.return_value.validate.return_value["groupid"] = (
+            sentinel.groupid
+        )
+        group_service.fetch.return_value = sentinel.duplicate_group
 
-        res = views.upsert(context, pyramid_request)
+        with pytest.raises(
+            HTTPConflict, match="group with groupid 'sentinel.groupid' already exists"
+        ):
+            views.update(context, pyramid_request)
 
-        groups_create.assert_called_once_with(pyramid_request)
-        assert res == groups_create.return_value
+        group_update_service.update.assert_not_called()
 
-    def test_it_does_not_proxy_to_create_if_group_extant(
-        self, context, pyramid_request, groups_create
-    ):
-        views.upsert(context, pyramid_request)
+    @pytest.fixture
+    def pyramid_request(self, pyramid_request):
+        pyramid_request.json_body = sentinel.json_body
+        return pyramid_request
 
-        assert not groups_create.call_count
 
-    def test_it_validates_against_group_update_schema_if_group_extant(
+@pytest.mark.usefixtures("group_service", "group_update_service")
+class TestUpsert:
```

Comment:
The `create()`, `update()` and `upsert()` views are quite similar, which leads to a lot of similar test code. Probably the view code could be refactored to extract some shared helper or service code, then we wouldn't need so many similar view tests. We should also [probably just remove the entire upsert API](https://hypothes-is.slack.com/archives/C4K6M7P5E/p1724926065440579).